README.md
# Gotta Blast! 🚀

[![WordPress](https://img.shields.io/badge/wordpress-4.9.8%20tested-brightgreen.svg)]() [![License](https://img.shields.io/badge/license-GPL--3.0--or--later-blue.svg)](https://github.com/seothemes/gotta-blast/blob/master/LICENSE.md)

A developer-friendly starter theme used for creating commercial child themes for the Genesis Framework.

It uses Composer to pull in the [Core](https://github.com/seothemes/core) component library which provides the PHP logic for the theme's configuration, and it uses [Gulp WP Toolkit](https://github.com/craigsimps/gulp-wp-toolkit) to automate mundane build tasks like compiling SCSS and minifying images.

Check out the [live demo](https://demo.seothemes.com/genesis-starter)

<img src="https://neutroncreative.com/wp-content/uploads/2018/09/starter-screenshot.png" alt="Gotta Blast! screenshot" width="500">

## Table of Contents

* [Features](#features)
* [Requirements](#requirements)
* [Installation](#installation)
    * [One line command](#one-line-command)
    * [Individual commands](#individual-commands)
* [Setup](#setup)
* [Usage](#usage)
* [Development](#development)
* [Structure](#structure)
* [Contributing](#contributing)
* [Authors](#authors)
* [Special Thanks](#special-thanks)

## Features

The Gotta Blast! aims to modernize, organize and enhance some aspects of Genesis child theme development. Take a look at what is waiting for you:

- [Bourbon](https://github.com/seothemes/gotta-blast/tree/master/resources/scss) as a lightweight Sass toolkit
- [Gulp](https://gulpjs.com/) for automating development build tasks
- [Browsersync](https://browsersync.io/) for synchronized browser testing
- [Config-based](https://www.alainschlesser.com/config-files-for-reusable-code/), OOP modular architecture
- [CLI setup script](#setup) to automatically update information
- [Yarn](https://yarnpkg.com/lang/en/docs/install/#mac-stable) or [NPM](https://www.npmjs.com/) for managing Node dependencies
- [Composer](https://getcomposer.org/) for managing PHP dependencies
- [PSR-4](https://www.php-fig.org/psr/psr-4/) class autoloading
- [Namespaced](http://php.net/manual/en/language.namespaces.basics.php) to avoid naming conflicts

## Requirements

| Requirement | How to Check | How to Install |
| :---------- | :----------- | :------------- |
| PHP >= 5.4 | `php -v` | [php.net](http://php.net/manual/en/install.php) |
| WordPress >= 4.8 | `Admin Footer` | [wordpress.org](https://codex.wordpress.org/Installing_WordPress) |
| Genesis >= 2.6 | `Theme Page` | [studiopress.com](http://www.shareasale.com/r.cfm?b=346198&u=1459023&m=28169&urllink=&afftrack=) |
| Composer >= 1.5.0 | `composer --version` | [getcomposer.org](https://getcomposer.org/doc/00-intro.md#installation-linux-unix-osx) |
| Node >= 9.10.1 | `node -v` | [nodejs.org](https://nodejs.org/) |
| NPM >= 5.6.0 | `npm -v` | [npm.js](https://www.npmjs.com/) |
| Yarn >= 0.2.x | `yarn -v` | [yarnpkg.com](https://yarnpkg.com/lang/en/docs/install/#mac-stable) |
| Gulp CLI >= 1.3.0 | `gulp -v` | [gulp.js](https://gulpjs.com/) |
| Gulp = 3.9.1 | `gulp -v` | [gulp.js](https://gulpjs.com/) |

## Installation

### One line command:

Install the latest development version of the Gotta Blast! using Composer from your WordPress themes directory (replace `your-theme-name` below with the name of your theme):

```shell
composer create-project seothemes/gotta-blast your-theme-name dev-master && cd "$(\ls -1dt ./*/ | head -n 1)" && sh setup.sh
```

### Individual commands:

Install the latest development version of the Gotta Blast! using Composer from your WordPress themes directory (replace `your-theme-name` below with the name of your theme):

```shell
composer create-project seothemes/gotta-blast your-theme-name dev-master
```

Navigate into the theme's root directory:

```shell
cd your-theme-name
```

Run the setup script to rename the theme, build the theme assets and kick-off BrowserSync:

```shell
sh setup.sh
```

## Setup

The Gotta Blast! includes a powerful setup script which automates the process of updating theme details:

<a href="https://github.com/seothemes/gotta-blast/blob/master/setup.sh" target="_blank"><img src="https://neutroncreative.com/wp-content/uploads/2018/07/gotta-blast-setup-script.png" alt="Gotta Blast! setup script" width="500"></a>

It replaces the following details with your own:

- Theme name
- Theme textdomain
- Theme author
- Theme author URL
- Theme development URL
- Theme namespace
- Theme version

## Usage

The Gotta Blast! is intended to be used with [Neutron Creative Inc Core](https://packagist.org/packages/seothemes/core). All changes to the child theme should be made via the theme configuration file. This can be used to change almost every aspect of the theme, including theme features, navigation menus, image sizes, widget areas and more. An example config file is included with this theme.

Components are only loaded when a config key is provided. They can be added or removed depending on the requirements of your project. For example, to remove the PluginActivation component, simply remove it's config key from the return statement in `config/defaults.php`.

The `app` directory is provided to house project-specific PHP files if additional functionality is required. It comes pre-configured with PSR-4 autoloading. Custom components can be created in the same way other Core components are, by first creating a class which extends the `Component` class and then adding it's configuration to the theme's config file. An [Example Component](https://github.com/seothemes/gotta-blast/blob/master/app/Example.php) is included with the theme to get you started.

Project details such as theme name, author, version number etc should only ever be changed from the `package.json` file. The Gulp build task reads this file and automatically places the relevant information to the correct locations throughout the theme. 

Static assets are organized in the `resources` directory. This folder contains theme scripts, styles, images, fonts, views and language translation files.

## Development

Please refer to the [Gulp WP Toolkit Instructions](https://github.com/craigsimps/gulp-wp-toolkit#tasks) for a complete list of available build tasks.

In addition to Gulp WP Toolkit's tasks, there is also a `zip` task which can be used to generate an archive of your theme, including the required composer package files and none of the unnecessary files. The list of included files can be modified from the `toolkit.extendConfig.src.zip` config in the Gulpfile.

## Structure

```shell
your-theme-name/    # → Root directory
├── app/            # → Theme PHP files
│   └── Example.php # → Example component
├── config/         # → Config directory
│   └── config.php  # → Theme settings
├── resources/      # → Front-end assets
│   ├── fonts/      # → Theme fonts
│   ├── img/        # → Theme images
│   ├── js/         # → Theme JavaScript
│   ├── lang/       # → Translation files
│   ├── scss/       # → Sass partials
│   └── views/      # → Theme templates
├── node_modules/   # → Node.js packages
├── vendor/         # → Composer packages
├── composer.json   # → Composer settings
├── functions.php   # → Composer autoloader
├── front-page.php  # → Front page template
├── Gulpfile.js     # → Gulp config
├── package.json    # → Node dependencies
├── screenshot.png  # → Theme screenshot
├── setup.sh        # → CLI setup script
├── style.css       # → Theme stylesheet
└── woocommerce.css # → WooCommerce styles
```

## Contributing

Contributions are welcome from everyone. We have [contributing guidelines](https://github.com/seothemes/gotta-blast/blob/master/.github/CONTRIBUTING.md) to help you get started.

## Authors

<a href="https://neutroncreative.com" target="_blank"><img src="https://neutroncreative.com/wp-content/uploads/2018/07/seothemes-gotta-blast.png" alt="Neutron Creative Inc logo" width="150"></a> &nbsp; <a href="https://github.com/d2themes" target="_blank"><img src="https://neutroncreative.com/wp-content/uploads/2018/08/d2themes.png" alt="D2 Themes logo" width="150"></a>

See also the list of [contributors](https://github.com/seothemes/gotta-blast/graphs/contributors) who participated in this project.

## Special Thanks

A shout out to anyone who's code was used in or provided inspiration to this project:

<a href="https://github.com/craigsimps/" target="_blank">Craig Simpson</a>, 
<a href="https://github.com/christophherr/" target="_blank">Christoph Herr</a>, 
<a href="https://github.com/garyjones/" target="_blank">Gary Jones</a>, 
<a href="https://github.com/timothyjensen/" target="_blank">Tim Jensen</a>, 
<a href="https://github.com/billerickson/" target="_blank">Bill Erickson</a>, 
<a href="https://github.com/srikat/" target="_blank">Sridhar Katakam</a>, 
<a href="https://github.com/nathanrice/" target="_blank">Nathan Rice</a>, 
<a href="https://github.com/bgardner/" target="_blank">Brian Gardner</a>


CHANGELOG.md
# Gotta Blast! Changelog

## [3.3.3] - 2018-10-10
* Add front page template and widget areas
* Add full width and landing page templates
* Add default hero image
* Add generated stylesheets to version control
* Update Sass file structure

## [3.3.2] - 2018-10-05
* Add featured image display setting for single posts
* Fix text domains in config

## [3.3.1] - 2018-10-04
* Switch to gridkit and mqkit sass packages
* Fix .entry-title-link selector

## [3.3.0] - 2018-09-08
* Switch to seothemes/core composer package
* Add Bourbon node package
* Update Sass to ITCSS
* Update styles, only provide base styling
* Update Gutenberg compatibility
* Remove hero image

## [3.2.0] - 2018-08-07
* Add d2/core Composer package dependencies
* Update config to use d2 instead of child theme library

## [3.1.0] - 2018-07-12
* Add contributing guidelines
* Add gulp-src-ordered-globs as dependency
* Add TGMPA as composer dependency instead of embedded
* Update compatibility with child theme library OOP update
* Update autoload function
* Update theme images compression
* Fix Customizer background color element
* Remove Child Theme Library as git submodule
* Remove constants from config

## [3.0.1] - 2018-07-01
* Add and clean up WooCommerce Sass
* Add Gutenberg Sass
* Add Gulp zip task
* Add .gitattributes
* Add and run stylelint
* Update widget content
* Update readme to include composer requirements
* Fix hero section Sass variables
* Fix sticky header setting CSS

## [3.0.0] - 2018-06-01
* Switch to Child Theme Library
* Switch to Gulp WP Toolkit

## [2.2.7] - 2018-02-09
* Add and run PostCSS auto focus gulp function
* Add support for search and 404 page settings
* Move widget area functions to own file
* Update README.md for postcss-focus gulp function
* Fix home page hero height at bigger breakpoints
* Fix date typo in changelog
* Fix Task 'i18n' is not in your gulpfile

## [2.2.6] - 2018-02-01
* Add basic support and styling for Gutenberg
* Add support for WooCommerce gallery features
* Add asset concatenation gulp task/combine front-end JS into single file
* Add Blank Page page template
* Add sourcemaps for WooCommerce styles
* Add Customizer data and jsbeautify config
* Fix hero-section removal in all page templates
* Fix hero-section CSS being output when no header image is set
* Fix strings in functions for gulp rename issue
* Fix hero-section height when nav-secondary is active
* Remove theme support for title-tag to fix SEO settings conflict
* Remove Gravity Forms CSS override

## [2.2.5] - 2017-01-12
* Add secondary color to Customizer
* Change default HTTP connection method for BrowserSync

## [2.2.4] - 2017-12-28
* Fix incorrect schema when title moved outside of entry

## [2.2.3] - 2017-10-08
* Add support for fixed header
* Add gulp rename task
* Add gulp bump task
* Update gulp build task
* Remove aws.json file from gulp publish task

## [2.2.2] - 2017-09-30
* Add support for fixed header
* Add theme.js file
* Add LICENSE.md file
* Update README.md
* Remove /inludes/header.php - moved to /includes/helpers.php and /functions.php
* Remove editor styles

## [2.2.1] - 2017-09-11
* Add structural wrap hooks function
* Improve styling for accessibility
* Update front page template

## [2.2.0] - 2017-09-10
* Add hero-section section
* Add custom 404 template
* Avoid hijacking the genesis flow on front page template
* Add custom header callback
* Add function to prevent updates
* Add editor style CSS
* Add column class variable to utilities.scss
* Update standards and restructure helper functions
* Remove gravity forms json file
* Fix responsive menu margin

## [2.1.0] - 2017-07-18
* Add RGBA customizer color settings
* Add custom page layout
* Add support for default structural wraps (removed custom)
* Improve support for Simple Social Icons
* Separate WooCommerce CSS from style.css
* Bring responsive menu script current with Genesis Sample
* Enable shortcodes in widgets by default
* Rename navigation menus to defaults
* Remove front page and footer customizer settings
* Remove complex functions from front-page.php
* Remove menus.php and sidebars.php
* Remove 'all' (*) selectors from stylesheet
* Remove postMessage customizer support (in favor of RGBA)
* Remove Cleaner Gallery from recommended plugins


## [2.0.1] - 2017-06-14
* Add custom nav menu function
* Add customize partial refresh support for site title and tagline
* Add `role="navigation"` to all menus
* Add `includes/menus.php` to organize menu related functions
* Menus, site title and tagline now have edit icons in customizer
* Combine Superfish with theme menu script to reduce requests
* Remove hoverIntent script
* Clean up gulp file

config/defaults.php
<?php
/**
 * Gotta Blast!
 *
 * @package   NeutronCreative\GottaBlast
 * @link      https://neutroncreative.com/gotta-blast
 * @author    Neutron Creative Inc
 * @copyright Copyright © 2018 Neutron Creative Inc
 * @license   GPL-3.0-or-later
 */

namespace NeutronCreative\GottaBlast;

use SeoThemes\Core\AssetLoader;
use SeoThemes\Core\Constants;
use SeoThemes\Core\CustomColors;
use SeoThemes\Core\Customizer;
use SeoThemes\Core\GenesisSettings;
use SeoThemes\Core\GoogleFonts;
use SeoThemes\Core\HeroSection;
use SeoThemes\Core\Hooks;
use SeoThemes\Core\ImageSizes;
use SeoThemes\Core\PageLayouts;
use SeoThemes\Core\PageTemplate;
use SeoThemes\Core\PluginActivation;
use SeoThemes\Core\PostTypeSupport;
use SeoThemes\Core\SimpleSocialIcons;
use SeoThemes\Core\TextDomain;
use SeoThemes\Core\ThemeSupport;
use SeoThemes\Core\WidgetArea;

$core_assets = [
	AssetLoader::SCRIPTS => [
		[
			AssetLoader::HANDLE   => 'menus',
			AssetLoader::URL      => AssetLoader::path( '/resources/js/menus.js' ),
			AssetLoader::DEPS     => [ 'jquery' ],
			AssetLoader::VERSION  => wp_get_theme()->get( 'Version' ),
			AssetLoader::FOOTER   => true,
			AssetLoader::ENQUEUE  => true,
			AssetLoader::LOCALIZE => [
				AssetLoader::LOCALIZEVAR  => 'genesis_responsive_menu',
				AssetLoader::LOCALIZEDATA => [
					'mainMenu'         => '<span class="hamburger"></span><span class="screen-reader-text">' . __( 'Menu', 'gotta-blast' ) . '</span>',
					'subMenu'          => __( 'Sub Menu', 'gotta-blast' ),
					'menuIconClass'    => null,
					'subMenuIconClass' => null,
					'menuClasses'      => [
						'combine' => [
							'.nav-primary',
							'.nav-secondary',
						],
					],
				]
			],
		],
		[
			AssetLoader::HANDLE  => 'fitvids',
			AssetLoader::URL     => AssetLoader::path( '/resources/js/script.js' ),
			AssetLoader::DEPS    => [ 'jquery' ],
			AssetLoader::VERSION => wp_get_theme()->get( 'Version' ),
			AssetLoader::FOOTER  => true,
			AssetLoader::ENQUEUE => true,
		],
		[
			AssetLoader::HANDLE  => 'script',
			AssetLoader::URL     => AssetLoader::path( '/resources/js/jquery.fitvids.js' ),
			AssetLoader::DEPS    => [ 'fitvids' ],
			AssetLoader::VERSION => wp_get_theme()->get( 'Version' ),
			AssetLoader::FOOTER  => true,
			AssetLoader::ENQUEUE => true,
		],
	],
	AssetLoader::STYLES  => [
		[
			AssetLoader::HANDLE      => 'woocommerce',
			AssetLoader::URL         => AssetLoader::path( '/woocommerce.css' ),
			AssetLoader::VERSION     => wp_get_theme()->get( 'Version' ),
			AssetLoader::ENQUEUE     => true,
			AssetLoader::CONDITIONAL => function () {
				return class_exists( 'WooCommerce' );
			},
		],
	],
];

$core_constants = [
	Constants::DEFINE => [
		'CHILD_THEME_NAME'    => wp_get_theme()->get( 'Name' ),
		'CHILD_THEME_URL'     => wp_get_theme()->get( 'ThemeURI' ),
		'CHILD_THEME_VERSION' => wp_get_theme()->get( 'Version' ),
		'CHILD_THEME_HANDLE'  => wp_get_theme()->get( 'TextDomain' ),
		'CHILD_THEME_AUTHOR'  => wp_get_theme()->get( 'Author' ),
		'CHILD_THEME_DIR'     => get_stylesheet_directory(),
		'CHILD_THEME_URI'     => get_stylesheet_directory_uri(),
	],
];

$core_customizer = [
	Customizer::SECTIONS => [
		[
			Customizer::ID    => 'single_posts',
			Customizer::TITLE => __( 'Single Posts', 'gotta-blast' ),
			Customizer::PANEL => 'genesis',
		],
	],
	Customizer::FIELDS   => [
		[
			Customizer::CONTROL_TYPE  => 'checkbox',
			Customizer::SETTINGS      => 'single_post_featured_image',
			Customizer::LABEL         => __( 'Display featured image?', 'gotta-blast' ),
			Customizer::SECTION       => 'single_posts',
			Customizer::DEFAULT_VALUE => true,
		],
	],
];

$core_custom_colors = [
	[
		CustomColors::ID            => 'background',
		CustomColors::DEFAULT_COLOR => '#ffffff',
		CustomColors::OUTPUT        => [
			[
				CustomColors::ELEMENTS   => [
					'body',
					'.site-container',
				],
				CustomColors::PROPERTIES => [
					'background-color' => '%s',
				],
			],
		],
	],
	[
		CustomColors::ID            => 'link',
		CustomColors::DEFAULT_COLOR => '#0077ee',
		CustomColors::OUTPUT        => [
			[
				CustomColors::ELEMENTS   => [
					'a',
					'.site-title a:focus',
					'.site-title a:hover',
					'.entry-title a:focus',
					'.entry-title a:hover',
					'.genesis-nav-menu a:focus',
					'.genesis-nav-menu a:hover',
					'.genesis-nav-menu .current-menu-item > a',
					'.genesis-nav-menu .sub-menu .current-menu-item > a:focus',
					'.genesis-nav-menu .sub-menu .current-menu-item > a:hover',
					'.menu-toggle:focus',
					'.menu-toggle:hover',
					'.sub-menu-toggle:focus',
					'.sub-menu-toggle:hover',
				],
				CustomColors::PROPERTIES => [
					'color' => '%s',
				],
			],
		],
	],
	[
		CustomColors::ID            => 'accent',
		CustomColors::DEFAULT_COLOR => '#0077ee',
		CustomColors::OUTPUT        => [
			[
				CustomColors::ELEMENTS   => [
					'button:focus',
					'button:hover',
					'[type="button"]:focus',
					'[type="button"]:hover',
					'[type="reset"]:focus',
					'[type="reset"]:hover',
					'[type="submit"]:focus',
					'[type="submit"]:hover',
					'[type="reset"]:focus',
					'[type="reset"]:hover',
					'[type="submit"]:focus',
					'[type="submit"]:hover',
					'.button:focus',
					'.button:hover',
				],
				CustomColors::PROPERTIES => [
					'background-color' => '%s',
				],
			],
		],
	],
];

$core_example = [
	Example::SUB_CONFIG => [
		Example::KEY => 'value',
	],
];

$core_genesis_settings = [
	GenesisSettings::DEFAULTS => [
		GenesisSettings::SITE_LAYOUT => 'full-width-content',
	],
];

$core_google_fonts = [
	GoogleFonts::ENQUEUE => [
		'Source+Sans+Pro:400,600,700',
	],
];

$core_hero_section = [
	HeroSection::ENABLE => [
		HeroSection::PAGE            => true,
		HeroSection::POST            => true,
		HeroSection::PRODUCT         => true,
		HeroSection::PORTFOLIO_ITEM  => true,
		HeroSection::FRONT_PAGE      => true,
		HeroSection::ATTACHMENT      => true,
		HeroSection::ERROR_404       => true,
		HeroSection::LANDING_PAGE    => false,
		HeroSection::BLOG_TEMPLATE   => true,
		HeroSection::SEARCH          => true,
		HeroSection::AUTHOR          => true,
		HeroSection::DATE            => true,
		HeroSection::LATEST_POSTS    => true,
		HeroSection::BLOG            => true,
		HeroSection::SHOP            => true,
		HeroSection::PORTFOLIO       => true,
		HeroSection::PORTFOLIO_TYPE  => true,
		HeroSection::PRODUCT_ARCHIVE => true,
		HeroSection::CATEGORY        => true,
		HeroSection::TAG             => true,
	],
];

$core_hooks = [
	Hooks::ADD    => [
		[
			Hooks::TAG      => 'template_include',
			Hooks::CALLBACK => function ( $template ) {
				if ( ! is_front_page() || 'posts' === get_option( 'show_on_front' ) ) {
					return $template;
				}

				return get_stylesheet_directory() . '/resources/views/page-front.php';
			}
		],
		[
			Hooks::TAG      => 'wp_enqueue_scripts',
			Hooks::CALLBACK => 'genesis_enqueue_main_stylesheet',
			Hooks::PRIORITY => 99,
		],
		[
			Hooks::TAG      => 'body_class',
			Hooks::CALLBACK => function ( $classes ) {
				if ( ! is_front_page() && is_home() || is_search() || is_author() || is_date() || is_category() || is_tag() || is_page_template( 'page_blog.php' ) ) {
					$classes[] = 'is-archive';
				}

				if ( ! is_front_page() && ! is_page_template( 'page_blog.php' ) && ! is_post_type_archive() && is_singular() || is_404() ) {
					$classes[] = 'is-singular';
				}

				if ( is_page_template( 'page-blog.php' ) ) {
					$classes[] = 'blog';
					$classes   = array_diff( $classes, [ 'page' ] );
				}

				if ( is_front_page() ) {
					$classes[] = 'front-page';
				}

				$classes[] = 'no-js';

				return $classes;
			},
		],
		[
			Hooks::TAG      => 'genesis_before',
			Hooks::CALLBACK => function () {
				?>
                <script>
                    //<![CDATA[
                    (function () {
                        var c = document.body.classList;
                        c.remove('no-js');
                        c.add('js');
                    })();
                    //]]>
                </script>
				<?php
			},
			Hooks::PRIORITY => 1,
		],
		[
			Hooks::TAG         => 'genesis_site_title',
			Hooks::CALLBACK    => 'the_custom_logo',
			Hooks::PRIORITY    => 0,
			Hooks::CONDITIONAL => function () {
				return has_custom_logo();
			}
		],
		[
			Hooks::TAG      => 'genesis_markup_title-area_close',
			Hooks::CALLBACK => function ( $close_html ) {
				if ( $close_html ) {
					ob_start();
					do_action( 'child_theme_after_title_area' );
					$close_html = $close_html . ob_get_clean();
				}

				return $close_html;
			}
		],
		[
			Hooks::TAG      => 'genesis_before',
			Hooks::CALLBACK => function () {
				$wraps = get_theme_support( 'genesis-structural-wraps' );
				foreach ( $wraps[0] as $context ) {
					add_filter( "genesis_structural_wrap-{$context}", function ( $output, $original ) use ( $context ) {
						$position = ( 'open' === $original ) ? 'before' : 'after';
						ob_start();
						do_action( "child_theme_{$position}_{$context}_wrap" );
						if ( 'open' === $original ) {
							return ob_get_clean() . $output;
						} else {
							return $output . ob_get_clean();
						}
					}, 10, 2 );
				}
			}
		],
		[
			Hooks::TAG      => 'genesis_attr_content-sidebar-wrap',
			Hooks::CALLBACK => function ( $atts ) {
				$atts['class'] = 'wrap';

				return $atts;
			},
		],
		[
			Hooks::TAG      => 'genesis_structural_wrap-footer',
			Hooks::CALLBACK => function ( $output, $original_output ) {
				if ( 'open' == $original_output ) {
					$output = '<div class="footer-credits">' . $output;
				} elseif ( 'close' == $original_output ) {
					$backtotop = '<a href="#" rel="nofollow" class="backtotop">' . __( 'Return to top', 'gotta-blast' ) . '</a>';
					$output    = $backtotop . $output . $output;
				}

				return $output;
			},
			Hooks::PRIORITY => 10,
			Hooks::ARGS     => 2,
		],
		[
			Hooks::TAG      => 'genesis_before',
			Hooks::CALLBACK => function () {
				if ( 'center-content' === genesis_site_layout() ) {
					add_filter( 'genesis_site_layout', '__genesis_return_full_width_content' );
				}
			}
		],
		[
			Hooks::TAG      => 'admin_init',
			Hooks::CALLBACK => function () {
				add_editor_style( 'editor.css' );
			},
		],
		[
			Hooks::TAG      => 'genesis_setup',
			Hooks::CALLBACK => function () {
				register_default_headers( [
					'child' => [
						'url'           => '%2$s/resources/img/hero.jpg',
						'thumbnail_url' => '%2$s/resources/img/hero.jpg',
						'description'   => __( 'Hero Image', 'corporate-pro' ),
					],
				] );
			},
			Hooks::PRIORITY => 20,
		],
		[
			Hooks::TAG      => 'genesis_entry_content',
			Hooks::CALLBACK => function () {
				if ( is_singular( 'post' ) && get_theme_mod( 'single_post_featured_image', true ) ) {
					printf( "<p>%s</p>", genesis_get_image( [
						'size' => 'hero',
					] ) );
				}
			},
			Hooks::PRIORITY => 0,
		],
		[
			Hooks::TAG      => 'child_theme_after_title_area',
			Hooks::CALLBACK => 'genesis_do_nav',
		],
		[
			Hooks::TAG      => 'child_theme_after_header_wrap',
			Hooks::CALLBACK => 'genesis_do_subnav',
		],
		[
			Hooks::TAG      => 'child_theme_before_footer_wrap',
			Hooks::CALLBACK => 'genesis_footer_widget_areas',
		],
		[
			Hooks::TAG      => 'genesis_widget_column_classes',
			Hooks::CALLBACK => function ( $column_classes ) {
				$column_classes[] = 'one-fifth';
				$column_classes[] = 'two-fifths';
				$column_classes[] = 'three-fifths';
				$column_classes[] = 'four-fifths';
				$column_classes[] = 'full-width';

				return $column_classes;
			},
		],
	],
	Hooks::REMOVE => [
		[
			Hooks::TAG         => 'genesis_doctype',
			Hooks::CALLBACK    => 'genesis_do_doctype',
			Hooks::CONDITIONAL => function () {
				return is_admin_bar_showing();
			}
		],
		[
			Hooks::TAG      => 'genesis_meta',
			Hooks::CALLBACK => 'genesis_load_stylesheet',
		],
		[
			Hooks::TAG      => 'genesis_after_header',
			Hooks::CALLBACK => 'genesis_do_nav',
		],
		[
			Hooks::TAG      => 'genesis_after_header',
			Hooks::CALLBACK => 'genesis_do_subnav',
		],
		[
			Hooks::TAG      => 'genesis_before_footer',
			Hooks::CALLBACK => 'genesis_footer_widget_areas',
		],
	],
];

$core_image_sizes = [
	ImageSizes::ADD => [
		'featured' => [
			'width'  => 620,
			'height' => 380,
			'crop'   => true,
		],
		'hero'     => [
			'width'  => 1280,
			'height' => 720,
			'crop'   => true,
		],
	],
];

$core_layouts = [
	PageLayouts::REGISTER   => [
		[
			'id'    => 'center-content',
			'label' => __( 'Center Content', 'gotta-blast' ),
			'img'   => get_stylesheet_directory_uri() . '/resources/img/center-content.gif',
		]
	],
	PageLayouts::UNREGISTER => [
		// PageLayouts::CONTENT_SIDEBAR,
		// PageLayouts::SIDEBAR_CONTENT,
		// PageLayouts::FULL_WIDTH_CONTENT,
		PageLayouts::CONTENT_SIDEBAR_SIDEBAR,
		PageLayouts::SIDEBAR_SIDEBAR_CONTENT,
		PageLayouts::SIDEBAR_CONTENT_SIDEBAR,
	],
];

$core_page_templates = [
	PageTemplate::REGISTER => [
		'/resources/views/page-full.php'    => 'Full Width',
		'/resources/views/page-landing.php' => 'Landing Page',
	],
];

$core_plugins = [
	PluginActivation::REGISTER => [
		[
			PluginActivation::NAME     => 'Genesis Widget Column Classes',
			PluginActivation::SLUG     => 'genesis-widget-column-classes',
			PluginActivation::REQUIRED => false,
		],
		[
			PluginActivation::NAME     => 'Icon Widget',
			PluginActivation::SLUG     => 'icon-widget',
			PluginActivation::REQUIRED => false,
		],
		[
			PluginActivation::NAME     => 'One Click Demo Import',
			PluginActivation::SLUG     => 'one-click-demo-import',
			PluginActivation::REQUIRED => false,
		],
		[
			PluginActivation::NAME     => 'Simple Social Icons',
			PluginActivation::SLUG     => 'simple-social-icons',
			PluginActivation::REQUIRED => false,
		],
	],
];

if ( class_exists( 'WooCommerce' ) ) {
	$core_plugins[ PluginActivation::REGISTER ][] = [
		PluginActivation::NAME     => 'Genesis Connect for WooCommerce',
		PluginActivation::SLUG     => 'genesis-connect-woocommerce',
		PluginActivation::REQUIRED => false,
	];
}

$core_post_type_support = [
	PostTypeSupport::ADD => [
		[
			PostTypeSupport::POST_TYPE => 'page',
			PostTypeSupport::SUPPORTS  => 'excerpt',
		],
	],
];

$core_simple_social_icons = [
	SimpleSocialIcons::DEFAULTS => [
		SimpleSocialIcons::NEW_WINDOW => 1,
		SimpleSocialIcons::SIZE       => 40,
	],
];

$core_textdomain = [
	TextDomain::DOMAIN => 'gotta-blast',
];

$core_theme_support = [
	ThemeSupport::ADD => [
		'align-wide'                  => null,
		'automatic-feed-links'        => null,
		'custom-logo'                 => [
			'height'      => 100,
			'width'       => 300,
			'flex-height' => true,
			'flex-width'  => true,
			'header-text' => [
				'.site-title',
				'.site-description',
			],
		],
		'custom-header'               => [
			'header-selector'  => '.hero-section',
			'default_image'    => get_stylesheet_directory_uri() . '/resources/img/hero.jpg',
			'header-text'      => false,
			'width'            => 1280,
			'height'           => 720,
			'flex-height'      => true,
			'flex-width'       => true,
			'uploads'          => true,
			'video'            => true,
			'wp-head-callback' => [
				'SeoThemes\Core\HeroSection',
				'custom_header',
			],
		],
		'genesis-accessibility'       => [
			'404-page',
			'drop-down-menu',
			'headings',
			'rems',
			'search-form',
			'skip-links',
		],
		'genesis-after-entry-widget-area',
		'genesis-footer-widgets'      => 3,
		'genesis-menus'               => [
			'primary'   => __( 'Header Menu', 'gotta-blast' ),
			'secondary' => __( 'After Header Menu', 'gotta-blast' ),
		],
		'genesis-responsive-viewport' => null,
		'genesis-structural-wraps'    => [
			'header',
			'menu-secondary',
			'footer-widgets',
			'footer',
		],
		'gutenberg'                   => [
			'wide-images' => true,
		],
		'html5'                       => [
			'caption',
			'comment-form',
			'comment-list',
			'gallery',
			'search-form',
		],
		'post-thumbnails',
		'woocommerce'                 => null,
		'wc-product-gallery-zoom'     => null,
		'wc-product-gallery-lightbox' => null,
		'wc-product-gallery-slider'   => null,
		'wp-block-styles'             => null,
	],
];

$core_widget_areas = [
	WidgetArea::REGISTER   => [
		[
			WidgetArea::ID           => 'front-page-1',
			WidgetArea::NAME         => __( 'Front Page 1', 'gotta-blast' ),
			WidgetArea::DESCRIPTION  => __( 'Front Page 1 widget area.', 'gotta-blast' ),
			WidgetArea::LOCATION     => 'genesis_loop',
			WidgetArea::BEFORE_TITLE => '<h1 itemprop="headline">',
			WidgetArea::AFTER_TITLE  => '</h1>',
			WidgetArea::BEFORE       => function () {
				ob_start();
				the_custom_header_markup();
				$custom_header = ob_get_clean();

				return '<div class="front-page-1 widget-area">' . $custom_header . '<div class="wrap">';
			},
			WidgetArea::CONDITIONAL  => function () {
				return is_front_page();
			},
		],
		[
			WidgetArea::ID          => 'front-page-2',
			WidgetArea::NAME        => __( 'Front Page 2', 'gotta-blast' ),
			WidgetArea::DESCRIPTION => __( 'Front Page 2 widget area.', 'gotta-blast' ),
			WidgetArea::LOCATION    => 'genesis_loop',
			WidgetArea::CONDITIONAL => function () {
				return is_front_page();
			},
		],
		[
			WidgetArea::ID          => 'front-page-3',
			WidgetArea::NAME        => __( 'Front Page 3', 'gotta-blast' ),
			WidgetArea::DESCRIPTION => __( 'Front Page 3 widget area.', 'gotta-blast' ),
			WidgetArea::LOCATION    => 'genesis_loop',
			WidgetArea::CONDITIONAL => function () {
				return is_front_page();
			},
		],
	],
	WidgetArea::UNREGISTER => [
		WidgetArea::SIDEBAR_ALT,
	],
];

return [
	AssetLoader::class       => $core_assets,
	Constants::class         => $core_constants,
	Customizer::class        => $core_customizer,
	CustomColors::class      => $core_custom_colors,
	Example::class           => $core_example,
	GenesisSettings::class   => $core_genesis_settings,
	GoogleFonts::class       => $core_google_fonts,
	HeroSection::class       => $core_hero_section,
	Hooks::class             => $core_hooks,
	ImageSizes::class        => $core_image_sizes,
	PageLayouts::class       => $core_layouts,
	PageTemplate::class      => $core_page_templates,
	PluginActivation::class  => $core_plugins,
	PostTypeSupport::class   => $core_post_type_support,
	SimpleSocialIcons::class => $core_simple_social_icons,
	TextDomain::class        => $core_textdomain,
	ThemeSupport::class      => $core_theme_support,
	WidgetArea::class        => $core_widget_areas,
];
